### PR TITLE
GO: Simplify Type Read/Write Generation

### PR DESCRIPTION
**Pull request by natewoods-wf**
### Background
Through performance profiling of our application, we found the frugal generated go code to take up around half the size of our transpiled javascript bundle.

### Description
This change inlines the `readFieldXXX` and `writeFieldXXX` method bodies into their respective `Read` and `Write` methods.  We fallback to generating `writeFieldXXX` for `set`s, `map`s and `list`s as abstracting the common logic for these would require more function invocations than the current implementation.  Inlining reduces total code and eliminates a method invocation for every field on reads.  We put the commonly generated three invocations of `writeFieldBegin`, `writeXXX`, and `writeFieldEnd` for the common base types.  This significantly reduces the amount of generated code, but does not reduce the number of method invocations when writing. 

### TL;DR
Changing how we generate Type Read/Writes shaves 3MB of data off our generated gopherjs bundle.

### Reviewers
@brianshannan-wf @Workiva/messaging-pp @Workiva/product2

### CODE REVIEW

**Comment by aviary-wf** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://dev.workiva.net/docs/teams/information-security/aviary#configuring-aviary) coverage of security relevant code
---
_Questions or Comments? Reach out on HipChat: InfoSec Forum._

File: examples/go/gen-go/v1/music/f_store_service.go [hunk 1]
```diff
@@ -475,33 +479,15 @@ func (p *StoreBuyAlbumArgs) Read(iprot thrift.TProtocol) error {
 	return nil
 }
 
-func (p *StoreBuyAlbumArgs) ReadField1(iprot thrift.TProtocol) error {
```
Comment (brianshannan-wf) [inline examples/go/gen-go/v1/music/f_store_service.go]: So removing this is technically a breaking change. No one should be using it, but from past experience, someone is probably using this method

Comment (natewoods-wf) [reply examples/go/gen-go/v1/music/f_store_service.go]: Looks like we are pretty safe: https://github.com/search?q=org%3AWorkiva+ReadField&type=Code

Comment (natewoods-wf) [reply examples/go/gen-go/v1/music/f_store_service.go]: Looks like usage across all of github is generated files or thrift specific: https://github.com/search?l=go&q=ReadField1&type=Code

Comment (brianshannan-wf) [reply examples/go/gen-go/v1/music/f_store_service.go]: While I do sympathize that these methods shouldn't have been publicly exposed and shouldn't be used anywhere, I think making a breaking change like this isn't something we want to do

Comment (brianshannan-wf) [reply examples/go/gen-go/v1/music/f_store_service.go]: If we do need to go forward with this, I would feel much better if this was implemented as an opt-in generation option until we are ready for a 3.0 release

File: compiler/generator/golang/generator.go [hunk 2]
```diff
@@ -1030,6 +1083,11 @@ func (g *Generator) GenerateTypesImports(file *os.File) error {
 	} else {
 		contents += "\t\"git.apache.org/thrift.git/lib/go/thrift\"\n"
 	}
+	if g.Options[frugalImportOption] != "" {
+		contents += "\t\"" + g.Options[frugalImportOption] + "\"\n"
+	} else {
+		contents += "\tfrugal \"github.com/Workiva/frugal/lib/go\"\n"
```
Comment (brianshannan-wf) [inline compiler/generator/golang/generator.go]: This import doesn't need an alias

File: compiler/generator/golang/generator.go [hunk 3]
```diff
@@ -1106,7 +1164,7 @@ func (g *Generator) GenerateServiceImports(file *os.File, s *parser.Service) err
 	if g.Options[frugalImportOption] != "" {
 		imports += "\t\"" + g.Options[frugalImportOption] + "\"\n"
 	} else {
-		imports += "\t\"github.com/Workiva/frugal/lib/go\"\n"
+		imports += "\tfrugal \"github.com/Workiva/frugal/lib/go\"\n"
```
Comment (brianshannan-wf) [inline compiler/generator/golang/generator.go]: This import doesn't need an alias

File: compiler/generator/golang/generator.go [hunk 4]
```diff
@@ -2156,6 +2214,33 @@ func (g *Generator) getGoTypeFromThriftTypePtr(t *parser.Type, pointer bool) str
 	}
 }
 
+func (g *Generator) getGoTypeFromThriftTypeEnum(typ *parser.Type) string {
```
Comment (brianshannan-wf) [inline compiler/generator/golang/generator.go]: I think I recognize this from somewhere else in the code, can this function be used there as well, or possibly combined with the type functions above?

File: lib/go/encoder.go [hunk 5]
```diff
@@ -0,0 +1,129 @@
+package frugal
+
+import "git.apache.org/thrift.git/lib/go/thrift"
+
+// WriteString writes string `value` of field name and id `name` and `field` respectively into `p`.
+func WriteString(p thrift.TProtocol, value, name string, field int16) error {
+	if err := p.WriteFieldBegin(name, thrift.STRING, field); err != nil {
+		return thrift.PrependError("write field begin error: ", err)
```
Comment (brianshannan-wf) [inline lib/go/encoder.go]: It would be good to get the field id and name in these error messages, so it's easier to tell what went wrong when errors are received

File: compiler/generator/golang/generator.go (hunk 4 above)
Comment (natewoods-wf) [reply compiler/generator/golang/generator.go]: The similar usage is in `generateWriteFieldRec` but also wrap it with the thrift protocol reading methods.  This is just the core go type.

File: lib/go/encoder.go (hunk 5 above)
Comment (natewoods-wf) [reply lib/go/encoder.go]: This responsibility is left to the caller, see these lines of `generateWriteFieldInline`

```go
	contents += indent + fmt.Sprintf("\tif err := frugal.Write%s(oprot, %s, \"%s\", %d); err != nil {\n", writeMethod, structField, field.Name, field.ID)
	contents += indent + fmt.Sprintf("\t\treturn thrift.PrependError(fmt.Sprintf(\"%%T::%s:%d \", p), err)", field.Name, field.ID)
```

File: compiler/generator/golang/generator.go (hunk 4 above)
Comment (brianshannan-wf) [reply compiler/generator/golang/generator.go]: Ah, right

File: compiler/generator/golang/generator.go [hunk 6]
```diff
@@ -895,7 +898,57 @@ func (g *Generator) generateReadFieldRec(field *parser.Field, first bool) string
...
+
+func (g *Generator) generateWriteFieldInline(field *parser.Field) (contents string) {
+	if !g.skipStandaloneFieldHandler(field) {
+		return fmt.Sprintf("\tif err := p.writeField%d(oprot); err != nil {\n\t\treturn err\n\t}\n", field.ID)
+	}
+
+	// Check if this field is optional and add nil checks if we need them.
+	var indent string
+	var tail string
+	if field.Modifier == parser.Optional || g.isPointerField(field) {
```
Comment (brianshannan-wf) [inline compiler/generator/golang/generator.go]: Why include `g.isPointerField()`? In the current code it only has an `IsSet()` check if the field is optional, which I believe is the desired behaviour

Comment (brianshannan-wf) [reply compiler/generator/golang/generator.go]: https://github.com/Workiva/frugal/blob/master/compiler/generator/golang/generator.go#L904 for reference

Comment (natewoods-wf) [reply compiler/generator/golang/generator.go]: good catch, cary over from when I was trying to determine if the value was nil manually.

Comment (natewoods-wf) [reply compiler/generator/golang/generator.go]: With the new `WriteStruct` method, if the pointer isn't checked for nil, we will panic in serialization.  And we generate the `IsSet` check for pointer types too.

https://github.com/Workiva/frugal/blob/master/compiler/generator/golang/generator.go#L525

**Comment by brianshannan-wf** [comment]
+1

**Comment by benechols-wf** [comment]
+1

**Comment by brianshannan-wf** [comment]
+1
